proto/greenfield/virtualgroup/tx.proto

```diff
@@ -0,0 +1,158 @@
+  string primary_sp_address = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // family_id is the identifier for the virtual group's family.
+  uint32 family_id = 2;
+  // secondary_sp_id is a list of secondary storage provider IDs associated with the virtual group.
+  repeated uint32 secondary_sp_ids = 3;
+  // total_deposit is the total deposit amount required for the virtual group.
+  // The tokens needs deposited and the size of storage are correlated.
+  cosmos.base.v1beta1.Coin deposit = 4 [(gogoproto.nullable) = false];
+}
+
+message MsgCreateGlobalVirtualGroupResponse {}
+
+message MsgDeleteGlobalVirtualGroup {
+  option (cosmos.msg.v1.signer) = "primary_sp_address";
+
+  // primary_sp_address defines the account address of the primary storage provider who creates the global virtual group.
+  string primary_sp_address = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // global_virtual_group_id is the identifier of the global virtual group.
+  uint32 global_virtual_group_id = 2;
+}
+
+message MsgDeleteGlobalVirtualGroupResponse {}
+
+message MsgDeposit {
+  option (cosmos.msg.v1.signer) = "funding_address";
+
+  // funding_address is the account address from which the deposit is made.
+  // The funding address must correspond to the funding address of the primary storage provider of the global virtual group.
+  string funding_address = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // global_virtual_group_id is the identifier of the global virtual group.
+  uint32 global_virtual_group_id = 2;
+  // deposit is the amount of tokens being deposited for the global virtual group.
+  cosmos.base.v1beta1.Coin deposit = 3 [(gogoproto.nullable) = false];
+}
+
+message MsgDepositResponse {}
+
+message MsgWithdraw {
+  // option (cosmos.msg.v1.signer) specifies the signer of the message.
+  option (cosmos.msg.v1.signer) = "funding_address";
+
+  // funding_address is the account address for withdrawing funds.
+  string funding_address = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // global_virtual_group_id is the identifier of the global virtual group.
+  uint32 global_virtual_group_id = 2;
+  // withdraw is the amount of coins to be withdrawn.
+  // The amount needs to be smaller than stored_size * storage_staking_price
+  cosmos.base.v1beta1.Coin withdraw = 3 [(gogoproto.nullable) = false];
+}
+
+message MsgWithdrawResponse {}
+
+message MsgSwapOut {
+  option (cosmos.msg.v1.signer) = "operator_address";
+
+  // operator_address defines the account address of the storage provider initiating the swap out.
+  string operator_address = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // virtual_group_family_id is the identifier of the virtual group family.
+  // if it set to 0, it represents that the operator swap out as the primary storage provider
+  // it it set to non-zero, it represents that the operator swap out as the secondary storage provider.
```

Comment:
The meaning of the comment here seems to be wrong.

Reply:
what wrong? I look fine

Reply:
// it it set to non-zero, swap out as the `primary` storage provider?
```go
if req.GlobalVirtualGroupFamilyId == types.NoSpecifiedFamilyId {
		// if the family id is not specified, it means that the SP will swap out as a secondary SP.
		err := k.SwapOutAsSecondarySP(ctx, sp.Id, successorSP.Id, req.GlobalVirtualGroupIds)
		if err != nil {
			return nil, err
		}
	} else {
		// if the family id is specified, it means that the SP will swap out as a primary SP and the successor sp will
		// take over all the gvg of this family
		err := k.SwapOutAsPrimarySP(ctx, sp, successorSP, req.GlobalVirtualGroupFamilyId)
		if err != nil {
			return nil, err
		}
	}
```